contracts/Marketplace.sol

```diff
@@ -208,35 +290,44 @@ contract Marketplace is Ownable {
     function grantSubscription(bytes32 productId, uint subscriptionSeconds, address recipient) public whenNotHalted onlyProductOwner(productId){
         return _subscribe(productId, subscriptionSeconds, recipient, false);
     }
+
+
     function buyFor(bytes32 productId, uint subscriptionSeconds, address recipient)  public whenNotHalted {
         return _subscribe(productId, subscriptionSeconds, recipient, true);
     }
 
+
      /**
      * Purchases access to this stream for msg.sender.
      * If the address already has a valid subscription, extends the subscription by the given period.
      * @dev since v4.0: Notify the seller if the seller implements PurchaseListener interface
      */
     function buy(bytes32 productId, uint subscriptionSeconds) public whenNotHalted {
-        buyFor(productId,subscriptionSeconds,msg.sender);
+        emit SubscriptionImported(productId, msg.sender,123);
+        buyFor(productId,subscriptionSeconds, msg.sender);
     }
 
 
+
     /**
     * Transfer a valid subscription from msg.sender to a new address.
     * If the address already has a valid subscription, extends the subscription by the msg.sender's remaining period.
     */
     function transferSubscription(bytes32 productId, address newSubscriber) public whenNotHalted {
-        (Product storage product, TimeBasedSubscription storage sub) = _getSubscription(productId, msg.sender);
+        _importSubscriptionIfNeeded(productId, msg.sender);
+        (Product storage product, TimeBasedSubscription storage sub) = _getSubscriptionLocal(productId, msg.sender);
         require(_isValid(sub), "error_subscriptionNotValid");
         uint secondsLeft = sub.endTimestamp.sub(block.timestamp);
         TimeBasedSubscription storage newSub = product.subscriptions[newSubscriber];
         _addSubscription(product, newSubscriber, secondsLeft, newSub);
-        delete product.subscriptions[msg.sender];
+        product.subscriptions[msg.sender] = TimeBasedSubscription(1);
         emit SubscriptionTransferred(productId, msg.sender, newSubscriber, secondsLeft);
     }
 
-    function _getSubscription(bytes32 productId, address subscriber) internal view returns (Product storage p, TimeBasedSubscription storage s) {
+    /**
+    gets subscriptions info for this
```

Comment:
Missing a word? this.. marketplace?

Reply:
I thought I'd leave it to the reader's interpretation... 

I'll add something more descriptive